contentful/client/base.py

```diff
@@ -0,0 +1,387 @@
+        reuse_entries: bool = False,
+        timeout_s: int = 1,
+        proxy_host: str | None = None,
+        proxy_port: str | None = None,
+        proxy_username: str | None = None,
+        proxy_password: str | None = None,
+        max_rate_limit_retries: str | None = 1,
+        max_rate_limit_wait: int = 60,
+        max_include_resolution_depth: int = 20,
+        application_name: str | None = None,
+        application_version: str | None = None,
+        integration_name: str | None = None,
+        integration_version: str | None = None,
+    ):
+        self.space_id = space_id
+        self.access_token = access_token
+        self.api_url = api_url
+        self.api_version = api_version
+        self.default_locale = default_locale
+        self.environment = environment
+        self.https = https
+        self.authorization_as_header = authorization_as_header
+        self.raw_mode = raw_mode
+        self.gzip_encoded = gzip_encoded
+        self.raise_errors = raise_errors
+        self.content_type_cache = content_type_cache
+        self.reuse_entries = reuse_entries
+        self.timeout_s = timeout_s
+        self.proxy_host = proxy_host
+        self.proxy_port = proxy_port
+        self.proxy_username = proxy_username
+        self.proxy_password = proxy_password
+        self.max_rate_limit_retries = max_rate_limit_retries
+        self.max_rate_limit_wait = max_rate_limit_wait
+        self.max_include_resolution_depth = max_include_resolution_depth
+        self.application_name = application_name
+        self.application_version = application_version
+        self.integration_name = integration_name
+        self.integration_version = integration_version
+        self._validate_configuration()
+        self._headers = None
+        self._client_info = None
+        self._proxy_info = None
+        self._transport = None
+
+    def initialize(self):
+        raise NotImplementedError()
+
+    def teardown(self):
+        raise NotImplementedError()
+
+    def _get(self, url: str, query: QueryT | None = None):
+        """
+        Wrapper for the HTTP Request,
+        Rate Limit Backoff is handled here,
+        Responses are Processed with ResourceBuilder.
+        """
+        raise NotImplementedError()
+
+    def _cache_content_types(self):
+        """
+        Updates the Content Type Cache.
+        """
+
+        raise NotImplementedError()
+
+    @property
+    def client_info(self) -> ClientInfo:
+        if self._client_info is None:
+            self._client_info = self._get_client_info()
+
+        return self._client_info
+
+    @property
+    def headers(self) -> dict[str, str]:
+        if self._headers is None:
+            self._headers = self._request_headers()
+
+        return self._headers
+
+    @property
+    def proxy_info(self) -> abstract.ProxyInfo:
+        if self._proxy_info is None:
+            self._proxy_info = self._proxy_parameters()
+
+        return self._proxy_info
```

Comment:
These were re-created with every request to Contentful. We're now only generating this information once, which is much more efficient.